src/Cardano/Wallet/Primitive.hs

```diff
@@ -134,37 +185,40 @@ newtype Address = Address
 
 instance NFData Address
 
+-- | This abstraction exists to give us the ability to keep the wallet business
+-- logic agnostic to the address derivation and discovery mechanisms.
+--
+-- This is needed because two different address schemes lives on Cardano:
+--   - A hierarchical random scheme:
+--      rather 'custom' made, with several flaws; this is the original and now
+--      legacy address scheme.
+--
+--   - A hierarchical sequential scheme:
+--      a new scheme based on the BIP-0044 specification, which is better suited
+--      for our present needs.
+--
+-- In practice, we will need a wallet that can support both, even if not at the
+-- same time, and this little abstraction can buy us this without introducing
+-- too much overhead.
+class IsOurs s where
+    isOurs :: Address -> s -> (Bool, s)
+
 
 -- * Coin
 
 -- | Coins are stored as Lovelace (reminder: 1 Lovelace = 1e6 ADA)
 newtype Coin = Coin
     { getCoin :: Word64
     } deriving stock (Show, Ord, Eq, Generic)
-      deriving newtype (Enum, Num, Real, Integral)
 
 instance NFData Coin
 
 instance Bounded Coin where
     minBound = Coin 0
     maxBound = Coin 45000000000000000
 
-instance Semigroup Coin where
-    (Coin a) <> (Coin b) =
-        invariant
-            ( mconcat
-                [ "Cardano.Wallet.Primitive.Coin (<>), sum out of bounds: "
-                , show a
-                , " + "
-                , show b
-                ]
-            )
-            (Coin (a + b))
-            (<= maxBound)
-
-instance Monoid Coin where
-    mempty  = minBound
-    mconcat = foldr (<>) mempty
+isValidCoin :: Coin -> Bool
+isValidCoin c = c >= minBound && c <= maxBound
```

Comment:
That's is mostly for testing now, and, generally speaking, since we don't have any smart-constructor on `Coin`, we are kinda free to put any value we want inside a `Coin`. So, in practice, we may need to verify this. This is dubious however, I admit.